--- a/madgraph/iolibs/template_files/makefile_sa_f2py	2017-08-17 00:25:24.000000001 +0200
+++ b/madgraph/iolibs/template_files/makefile_sa_f2py.patched	2018-05-05 14:30:53.000000001 +0200
@@ -14,7 +14,7 @@
 
 
 allmatrix$(MENUM)py.so: $(LIBDIR)/$(LIBRARY)  all_matrix.o
-	$(F2PY) $(LINKLIBS) -lmatrix  -c all_matrix.f */matrix.f -m allmatrix$(MENUM)py
+	$(F2PY) --f77exec=$(FC) $(LINKLIBS) -lmatrix  -c all_matrix.f */matrix.f -m allmatrix$(MENUM)py
 
 
 
